thavalon-server/src/game/engine.rs

```diff
@@ -0,0 +1,761 @@
+                                Entry::Vacant(entry) => {
+                                    let role = game.players[player].role;
+                                    if role.can_play(card) {
+                                        entry.insert(card);
+                                        log::debug!("{} played a {}", game.name(player), card);
+                                        Ok(())
+                                    } else {
+                                        Err(format!("You can't play a {}", card))
+                                    }
+                                }
+                            },
+                            // TODO: fun on-mission actions
+                            _ => Err("You can't do that right now".to_string()),
+                        }
+                    } else {
+                        Err("You are not on the mission".to_string())
+                    }
+                })
+                .await?;
+        }
+
+        let failed = is_failure(game.spec, mission, cards.values());
+        log::debug!(
+            "Mission {} {}",
+            mission,
+            if failed { "failed" } else { "passed" }
+        );
+
+        self.interactions
+            .send(Message::MissionResults {
+                passed: !failed,
+                successes: cards.values().filter(|c| c == &&Card::Success).count(),
+                fails: cards.values().filter(|c| c == &&Card::Fail).count(),
+                reverses: cards.values().filter(|c| c == &&Card::Reverse).count(),
+            })
+            .await?;
+
+        if !failed {
+            let agravaine_declared = match time::timeout(
+                DECLARE_DELAY,
+                self.interactions.receive(|player, action| {
+                    let role = game.players[player].role;
+                    match action {
+                        Action::Declare if role == Role::Agravaine => Ok(true),
+                        _ => Err("You can't do that right now".to_string()),
+                    }
+                }),
+            )
+            .await
+            {
+                Ok(Ok(declared)) => declared,
+                // We either timed out waiting for Agravaine to declare or encountered an error receiving messages. Since there won't
+                // necessarily _be_ a message in this time frame, ignore errors.
+                _ => false,
+            };
+
+            if agravaine_declared {
+                log::debug!("Agravaine declared!");
+
+                // maybe this should be a different message?
+                self.interactions
+                    .send(Message::MissionResults {
+                        passed: false,
+                        successes: cards.values().filter(|c| c == &&Card::Success).count(),
+                        fails: cards.values().filter(|c| c == &&Card::Fail).count(),
+                        reverses: cards.values().filter(|c| c == &&Card::Reverse).count(),
```

Comment:
If we cached these values from above while counting, we could reuse them here without needing to filter again.